### PR TITLE
add autogen code support for matrix_nms.

### PR DESCRIPTION
**Pull request by ScottWong98**
### PR types

Others

### PR changes

Others

### Describe

add autogen code support for `matrix_nms`.

#51842

### CODE REVIEW

File: paddle/phi/api/yaml/op_compat.yaml [hunk 1]
```diff
@@ -1237,6 +1237,12 @@
     attrs : [bool use_mkldnn = false, float scale_x = 1.0f, 'float[] scale_y = {1.0f}',
              float scale_out = 1.0f, bool force_fp32_output = false]
 
+- op : matrix_nms
+  inputs :
+    {bboxes : BBoxes, scores : Scores}
+  outputs :
+    {out : Out, index : Index, roisnum : RoisNum}
```
Comment (heavyrain-lzy) [inline paddle/phi/api/yaml/op_compat.yaml:1244]: 由于该算子的`GetExpectedKernelType`特殊，需要手动支持，这里建议如下处理这种情况：
1.这里增加配置
```
get_expected_kernel_type :
    matrix_nms : GetMatrixNmsExpectedKernelType
```
2.`paddle/fluid/operators/generator/get_expected_kernel_func.cc`文件增加函数
```
phi::KernelKey GetMatrixNmsExpectedKernelType(
      const framework::ExecutionContext& ctx,
      const framework::OperatorWithKernel* op_ptr) {
  return phi::KernelKey(
      op_ptr->IndicateVarDataType(ctx, "Scores"),
      platform::CPUPlace());
  }
```
3. 头文件`paddle/fluid/operators/generator/get_expected_kernel_func.h`增加函数声明
```
phi::KernelKey GetMatrixNmsExpectedKernelType(
      const framework::ExecutionContext& ctx,
      const framework::OperatorWithKernel* op_ptr);
```

File: paddle/phi/api/yaml/ops.yaml [hunk 2]
```diff
@@ -1001,6 +1001,14 @@
     data_type : x
   backward : masked_select_grad
 
+- op : matrix_nms
+  args : (Tensor bboxes, Tensor scores, float score_threshold, int nms_top_k, int keep_top_k, float post_threshold=0., bool use_gaussian = false, float gaussian_sigma = 2.0, int background_label = 0, bool normalized = true)
+  output : Tensor(out), Tensor(index), Tensor(roisnum)
+  infer_meta :
```
Comment (heavyrain-lzy) [inline paddle/phi/api/yaml/ops.yaml:1007]: 增加`optional : roisum`

File: paddle/phi/api/yaml/ops.yaml [hunk 3]
```diff
@@ -1001,6 +1001,14 @@
     data_type : x
   backward : masked_select_grad
 
+- op : matrix_nms
+  args : (Tensor bboxes, Tensor scores, float score_threshold, int nms_top_k, int keep_top_k, float post_threshold=0., bool use_gaussian = false, float gaussian_sigma = 2.0, int background_label = 0, bool normalized = true)
```
Comment (heavyrain-lzy) [inline paddle/phi/api/yaml/ops.yaml]: `gaussian_sigma = 2.0`修改为`gaussian_sigma = 2.`

File: paddle/phi/api/yaml/ops.yaml (hunk 2 above)
Comment (ScottWong98) [reply paddle/phi/api/yaml/ops.yaml:1007]: Done

File: paddle/phi/api/yaml/op_compat.yaml (hunk 1 above)
Comment (ScottWong98) [reply paddle/phi/api/yaml/op_compat.yaml:1244]: Done

File: paddle/phi/api/yaml/ops.yaml (hunk 3 above)
Comment (ScottWong98) [reply paddle/phi/api/yaml/ops.yaml]: Done

**Review by heavyrain-lzy — APPROVED** [review]
LGTM

**Pull request merged**